### ISSUE TITLE
awscloudwatchlogsexporter exporter needs to support dynamic log stream names

### ISSUE THREAD

**Issue by thomasbaldwin** [open-telemetry/opentelemetry-collector-contrib#10887]
**Describe the bug**
If you are running a daemonset with a configmap that uses the awscloudwatchlogsexporter plugin as an exporter, and you have more than one node, you will have an issue in which multiple actors are trying to write to the same log stream at the same time.  This causes an issue with cloud watch in which it will say you have an invalid sequence token.

**Steps to reproduce**
If possible, provide a recipe for reproducing the error.

**What did you expect to see?**
I would like to have something like the [awsemfexporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/awsemfexporter) has, in which you can provide `{NodeName}` as the log stream name, and it will resolve it to the `k8s.node.name`

**What did you see instead?**
No ability to do this

**What version did you use?**
0.51

**What config did you use?**
Config: (e.g. the yaml config file)

**Environment**
AWS EKS

**Comment by djaglowski** [comment]
@boostchicken, pinging you as code owner.

**Comment by boostchicken** [comment]
Sure let me give this a look and see if I can colme up with an LOE.

**Comment by boostchicken** [comment]
@thomasbaldwin can you provide some configs or more artifacts that cause this issue?  If I can easily reproduce it I am sure I can find a quick fix.

I am hesistant to just resolve k8s.node.name and do that for everyone.  Maybe a syntax for using any variable in the title which is a much bigger change.

@djaglowski who else should we involve in this discussion in terms of the overall direction that the collector is headed.  I don't want to add a bunch of functionality that is out of line with the project as a whole.

**Comment by djaglowski** [comment]
Thanks for the quick response @boostchicken. 

Looking at this a little closer, it's not clear to me that this is really an issue for this component to solve. If I'm understanding the issue correctly, the proposed idea ultimately is adding or modifying an attribute on the log record. Is that right? If so, it probably should be handled by a processor (eg `attributesprocessor` or `transformprocessor`).

**Comment by boostchicken** [comment]
Upon further thought, the real issue here is there is no singleton for exporting to AWS services.  A name just might solve it.  There is a sequence counter in the API call, if its out of order it is rejected, lots of services have it.

So I think what we need to do here is find every AWS API call in Otel-contrib and funnel it through a Singleton for each AWS Service Client.  Make a factory, then share instances across threads.

@djaglowski This is no small change, if I go about it I would still like people's blessing.  Also does AWS have anything to say about this?

**Comment by boostchicken** [comment]
So using a single client might fix it, but I must yield to the documented AWS solution

https://aws.amazon.com/premiumsupport/knowledge-center/cloudwatch-invalid-sequence-token/#:~:text=InvalidSequenceToken%20errors%20when%20multiple%20sources,stream%2C%20you%20receive%20an%20error.

We need to make the log streams unique.  Ill take a peak and see what the emf exporter does and if i cn make it generic.

**Comment by boostchicken** [comment]
OK found the code here it is

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/0c691295193854523240b609c0c3cc06f9202b41/exporter/awsemfexporter/util.go#L92-L113

We need to refactor all of that out to the internal/aws namespace, update awsemfexporter, and then update cloudwatchlogsexporter

The change could be done quicker if we duplicate code, I doubt anyone wants to do that

@djaglowski Gonna wait on comments before I move forward.

**Comment by boostchicken** [comment]
Yeah this is gonna be ugly, we gotta re-do the whole exporter.  It only creates 1 Pusher because it assumes the name never changes.  The name could change on any attribute that is on a Logs.Resource().  Attributes are not in scope till translation, the pusher has already been made.  Also, since we will now have multiple pushers we are going to have to have a Map of pushers so we don't re-create them all the time.

**Comment by boostchicken** [comment]
Paging @aateeqi

Is AWS willing to pickup this work or are we on our own?

**Comment by thomasbaldwin** [comment]
@boostchicken thanks so much for your quick replies! is there a quick and easy way you would recommend getting it done? Even if it's a little sloppy, this is a blocker for something I'm working on, so I would be willing to run my own image until this feature request gets merged in. I would also be willing to work with you on the change but might need a little bit of help getting up and running

**Comment by boostchicken** [comment]
I pretty much outlined it above, could make a quick copy paste job out of it.  I wanted to give AWS sometime to chime in, I suppose they aren't gonna support this.  If you wanna take a crack at it, go for it.  If not I think I can get to this in a couple days. 

Cheers!

**Comment by boostchicken** [comment]
The only thing that makes me hesitate is the need to refactor the functions I posted above out into the internal/aws space.  It's a pretty material change and I am not sure how it might impact tests etc.  

You could literally copy and paste these functions, replacing the current pusher and fix this.  That would be the ugly way which would not be of quality high enough to be considered for an OSS project imho.

**Comment by boostchicken** [comment]
Hey @thomasbaldwin give this a go https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/12179

**Comment by thomasbaldwin** [comment]
sounds good @boostchicken ! will check it out and post back!
